### PR TITLE
Feat: Support more constraint operators

### LINKED ISSUE #42: feat: Add support for new constraint operators
I have written a summary In this issue: https://github.com/Unleash/unleash/issues/856#issuecomment-1027910183

Client-specifications: https://github.com/Unleash/client-specification/pull/25

Node SDK reference implementation: https://github.com/Unleash/unleash-client-node/pull/289

**PS! we are still working out the UI and validating the actual protocol, but we are at > 93% certainty.**

### PR DESCRIPTION
**Pull request by RikudouSage**
closes #42 

Before merging the reference to client-specifications repository should be put back to the `main` branch.

### CODE REVIEW

File: tests/ConstraintValidator/Operator/AbstractOperatorValidatorTest.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                    'enabled' => true,
+                    'strategies' => [
+                        [
+                            'name' => 'default',
+                            'constraints' => [
+                                [
+                                    'contextName' => 'email',
+                                    'operator' => ConstraintOperator::STRING_STARTS_WITH,
+                                    'inverted' => true,
+                                    // missing 'value' and 'values' and 'inverted' set to true
```
Comment (ivarconr) [inline tests/ConstraintValidator/Operator/AbstractOperatorValidatorTest.php:98]: nice =)

File: src/ConstraintValidator/Operator/Number/AbstractNumberOperatorValidator.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+{
+    protected function acceptsValues(array|string $values): bool
+    {
+        return is_string($values) && is_numeric($values);
+    }
+
+    protected function convert(string $number): int|float
+    {
+        if (str_contains($number, '.')) {
+            return (float) $number;
```
Comment (ivarconr) [inline src/ConstraintValidator/Operator/Number/AbstractNumberOperatorValidator.php:20]: nice, I will add a float test to the specifications also.

File: src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * @internal
+ */
+final class StringEndsWithOperatorValidator extends AbstractStringOperatorValidator
+{
+    protected function validate(string $currentValue, array|string $searchInValue): bool
+    {
+        assert(is_string($searchInValue));
+
+        return str_ends_with($searchInValue, $currentValue);
```
Comment (ivarconr) [inline src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php:14]: Is `str_ends_with` supported in PHP 7.2+, or is it only PHP 8.0?

Comment (ivarconr) [reply src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php:14]: https://www.codegrepper.com/code-examples/php/str_ends_with+php+7

File: src/ConstraintValidator/Operator/String/StringStartsWithOperatorValidator.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * @internal
+ */
+final class StringStartsWithOperatorValidator extends AbstractStringOperatorValidator
+{
+    protected function validate(string $currentValue, array|string $searchInValue): bool
+    {
+        assert(is_string($searchInValue));
+
+        return str_starts_with($currentValue, $searchInValue);
```
Comment (ivarconr) [inline src/ConstraintValidator/Operator/String/StringStartsWithOperatorValidator.php:14]: only from PHP 8?

File: src/ConstraintValidator/Operator/String/StringContainsOperatorValidator.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * @internal
+ */
+final class StringContainsOperatorValidator extends AbstractStringOperatorValidator
+{
+    protected function validate(string $currentValue, array|string $searchInValue): bool
+    {
+        assert(is_string($searchInValue));
+
+        return str_contains($currentValue, $searchInValue);
```
Comment (ivarconr) [inline src/ConstraintValidator/Operator/String/StringContainsOperatorValidator.php:14]: only from PHP 8?

**Review by ivarconr — COMMENTED** [review]
It looks good! 

It looks to me some of the operators used will only work on PHP 8. It would be good to use simple work-around to make sure they also work on PHP 7, given that it is possible and they are straight forward.

File: src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php (hunk 3 above)
Comment (RikudouSage) [reply src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php:14]: 8.0, but this repo contains a downgrader which downgrades the code to multiple version, it gets automatically changed to something suitable for each version during deploy, see for example [here](https://github.com/Unleash/unleash-client-php/commit/f0a215569dba6260c44f56f88c30f42cb8cd0433) for version downgraded to php 7.4.

Comment (RikudouSage) [reply src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php:14]: And here you can see one of the str_* functions downgraded: https://github.com/Unleash/unleash-client-php/blob/f0a215569dba6260c44f56f88c30f42cb8cd0433/src/Configuration/UnleashConfiguration.php#L59

File: src/ConstraintValidator/Operator/String/StringStartsWithOperatorValidator.php (hunk 4 above)
Comment (RikudouSage) [reply src/ConstraintValidator/Operator/String/StringStartsWithOperatorValidator.php:14]: Explained in previous comments, gets downgraded to suitable expression for each version.

**Comment by RikudouSage** [comment]
As mentioned before, it gets downgraded for each version, that's why it's kinda easy to support older versions without sacrificing the features of new versions.

File: src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php (hunk 3 above)
Comment (ivarconr) [reply src/ConstraintValidator/Operator/String/StringEndsWithOperatorValidator.php:14]: This is nice. Thanks for explaining 👍🏼

**Review by ivarconr — APPROVED** [review]
lgtm :D

**Pull request merged**